lib/scala/cli/src/main/scala/org/enso/cli/Spelling.scala

```diff
@@ -0,0 +1,150 @@
+package org.enso.cli
+
+object Spelling {
+
+  /**
+    * Selects closest matches to a potentially mistyped name.
+    * @param typo the unknown name that the user typed
+    * @param possibleWords a sequence of words that were expected
+    * @return a sequence of `possibleWords` that are most similar to `typo`
+    */
+  def selectClosestMatches(
+    typo: String,
+    possibleWords: Seq[String]
+  ): Seq[String] =
+    selectClosestMatchesWithMetadata(typo, possibleWords.map((_, ()))).map(_._1)
+
+  /**
+    * Selects closest matches to a potentially mistyped name, returning not only
+    * the matches but any metadata that was passed with them. May be useful for
+    * including additional text in the match suggestions.
+    *
+    * @param typo the unknown name that the user typed
+    * @param possibleMatches a sequence of tuples (word, metadata)
+    * @tparam A type of metadata
+    * @return a sequence of tuples from `possibleMatches` whose first elements
+    *         were most similar to `typo`
+    */
+  def selectClosestMatchesWithMetadata[A](
+    typo: String,
+    possibleMatches: Seq[(String, A)]
+  ): Seq[(String, A)] =
+    if (
+      possibleMatches.isEmpty ||
+      typo.length > 2 * possibleMatches.map(_._1.length).max
+    ) Seq()
+    else {
+      val withScores =
+        possibleMatches.map(word => (scoreNeedlemanWunsch(typo, word._1), word))
+      val top = withScores
+        .filter(_._1 < maxScoreThreshold)
+        .sortBy(_._1)
+        .take(topSuggestionsLimit)
+      top.map(_._2)
+    }
+
+  /**
+    * A threshold for the Needleman-Wunsch similarity score. Any matches above
+    * that threshold are not considered.
+    */
+  private val maxScoreThreshold: Double = 10.0
+
+  /**
+    * The limit of how many suggestions should be displayed at most.
+    */
+  private val topSuggestionsLimit = 3
+
+  /**
+    * Computes the similarity score between two words using the Needleman-Wunsch
+    * algorithm, as described at
+    * https://en.wikipedia.org/wiki/Needleman%E2%80%93Wunsch_algorithm
+    *
+    * Uses approximate Euclidean distance between keys on the QWERTY keyboard as
+    * the mismatch score. The weights are chosen in such a way, that the smaller
+    * the score, the bigger is the similarity between the two words.
+    */
+  def scoreNeedlemanWunsch(wordA: String, wordB: String): Double = {
```

Comment:
It will be more clear if you create a separate object with the distance algorithm.
Also, it is worth adding a couple of tests covering either the distance algorithm or the final spell-checking function (or both). It will help to maintain the module in the future.

Reply:
Added a test for the `selectClosestMatches` function. The score itself is highly heuristics, so its tests would be rather brittle. But the `selectClosestMatches` test ensures that it works reasonably well.